### ISSUE TITLE
Transition to using min_const_generics for user-facing type-level numbers

### ISSUE THREAD

**Issue by kwf** [boltlabs-inc/dialectic#58]
The `min_const_generics` feature of Rust stabilizes on ~~March 25~~ some unknown date. This feature enables us to do away with the constants defined in `dialectic::unary::constants` and `dialectic::unary::types`, and instead use const generics to specify numbers in the user interface. Because const generics in the minimal formulation currently present do not do symbolic computation (i.e. compute on unknown constants), we still need to internally convert to unary numbers, but we can hide this from the user.

This change involves:

- The signature of `Chan::choose` will now take a turbofish-specified const generic argument rather than a value-level ZST representing a unary number, leading to syntax like `chan.choose::<1>()` over the previous `chan.choose(_1)`.
- The surface syntax of the `offer!` macro will take normal integer literals in its arms, not underscore-prefixed ones.
- The surface syntax of the `choose` and `offer` constructs in the `Session!` macro will also take normal integer literals in their arms, not underscore-prefixed ones.
- The de Bruijn index type argument of `Continue` will become a const generic argument rather than a unary number.
- The upper bound type argument of `Choice` will become a const generic argument rather than a unary number.
- The modules `dialectic::unary::types` and `dialectic::unary::constants` will entirely be removed.
- A new sealed trait, `ToUnary`, will be introduced, which has a single associated type with a `Unary` bound. Each constant from `0` to `128` will be given an implementation of this trait (we can do more, but that's all we need for now), auto-generated by a new procedural macro exposed by the dialectic-macro crate.
- The various bounds requiring arguments which have changed per above from unary numbers to const generics will be updated from `Unary` to `ToUnary`, and various types referring to those unary numbers will need to be updated to use the converted unary number rather than the type parameter directly.
- Documentation and tests, including the tutorial, item documentation, quick reference, and examples, will need to be updated to reflect the new interface.

From how I see it, this breaks down into several phases, in this order:

0. Switch our development process temporarily to using the beta compiler, including updating our CI infrastructure to use the beta channel. This will be reverted once `min_const_generics` lands on stable.
1. Change the surface syntax of `Session!` and `offer!` to use non-underscore-prefixed numerals, and alter their documentation and associated tests to match (including tutorial).
2. Define the `ToUnary` trait and the procedural macro to generate all its instances. Document and test.
3. Change the generic parameter of `Continue` to use a const generic argument, altering all necessary traits in the `types` module to depend on `ToUnary` for indexing when doing substitution, scope-checking, etc. Also alter the `Session!` macro to generate numerals rather than unary numbers as the argument to `Continue` in the `ToTokens` implementation of `Target`. Update documentation and tests to reflect this change. (This change should not require modifying the examples, since they should transparently continue to work.)
4. Change the generic parameter of `Choice` to use a const generic parameter.
5. Change the generic parameter of `Chan::choose` to take a const generic argument, removing its value level argument. Alter its documentation (both for the item and in the tutorial and quick reference) and update its tests, including in examples.
6. Remove the `unary::types` and `unary::constants` modules, and ensure that no documentation, tests, or code continue to refer to them (double-check the tutorial).

Ordering constraints:

- 0 must happen before all others.
- 1 and 2 can happen simultaneously.
- 3, 4, and 5 can happen simultaneously.
- 6 must happen after all others.

I propose that we create a new development branch for this change, and make each change as a separate pull request against that development branch.

**Comment by kwf** [comment]
Regarding the `ToUnary` trait, this is how it seems like it must be defined. There must be a wrapper type `Number` to convert from the land of constants to the land of types; you can't directly implement a trait on a constant:

```rust
pub trait ToUnary {
    type AsUnary: Unary;
}

struct Number<const N: usize>(PhantomData<[(); N]>);

impl ToUnary for Number<0> {
    type AsUnary = Z;
}
```

**Comment by kwf** [comment]
Also, seems like the timeline for `min_const_generics` stabilization is farther out than we thought? Need more info about that.

**Comment by sdleffler** [comment]
As we've confirmed that min_const_generics will hit stable on 2021-03-25, it seems likely that we'll be preparing for a release 0.4 on or slightly after 2021-03-25, using the new feature to improve the user-facing API. Since we're now on board this conversion, there's an important question that needs to be answered: should `Continue<N>` have `N` be a const generic, or be a unary type? If we make it a unary type, then we can leave codegen in the macro compiler unchanged; if we instead have a `Continue_<N>` or `UnaryContinue<N>` which takes an `N: Unary` rather than `const N: usize`, and then have a type synonym `pub type Continue<N> = Continue_<<Number<N> as ToUnary>::AsUnary>;` then the user-facing API can nicely use `Continue<1>`, `Continue<0>`, etc. while having the unary continue exist under the hood.

**Comment by kwf** [comment]
Choosing to leave it as unary means that error messages will mention fully-expanded unary numbers, which I think is less intelligible than mentioning constants. Given that we are merely emitting a token stream, I don't think it should be a problem to emit a numeric token in the correct position, should it?

**Comment by sdleffler** [comment]
This is true, but if we emit unary type level integers we could avoid modifying the codegen significantly and also possibly speed up compilation a tiny bit by not putting the onus on Rust to convert the constants into type-level unaries.

**Comment by sdleffler** [comment]
On second thought, I don't think the performance gain from that would be significant. We could find out, but for now I'll emit integers and swap `Continue` completely over to using const usize.

**Comment by sdleffler** [comment]
Actually, it gets worse. `Continue` has type-level addition/similar operations performed directly on it. We could convert const generics to unary and then perform the ops and then go back with *another* trait conversion, but this is all starting to feel a lot like basically just bandaging over the fact that const generics are not ready for arbitrary usage like this yet. We can totally use them for user facing stuff but once we convert to unary, we really don't want to go back if we can avoid it, I think.

**Comment by kwf** [comment]
Could we not relatively easily generate impls for a `FromUnary` trait alongside those for the `ToUnary` trait, just as we need to generate both `List` and `Tuple` to get both directions for tuple-to-list conversions?

Another note: we shouldn't forget to seal the `ToUnary` (and if implemented, `FromUnary`) traits.

**Comment by sdleffler** [comment]
So, we can, but there's a couple glitches in that; at the type level, we cannot go from some type containing a `const N: usize` without binding it to a type variable, and then "solving" for it. The main issue with this is that it is not possible to uniquely specify a const generic parameter without having it in the type being implemented for (trait or non-trait impl.) Const generics *cannot be produced from type families*. This is annoying.

One possible workaround is not to have `Continue<const N: usize>` but `Continue<Number<N>>`, and since `Number<N>` is of the type kind rather than the const generic `usize` kind, we can treat it w/ a type family (and we could also potentially have an `Indexed` trait that produces a `Number<const N: usize>` from a type implementing it.) I would just like to note that *every* time we end up doing something that modifies a `Continue`, we'll end up going from const to `Unary` to const again. That feels very hacky but I guess it's just the machinery needed to work around the current drawbacks of const generics.

**Comment by kwf** [comment]
Perhaps we can defer this to last, since these modifications are independent of all others? I think I have some ideas of how to have and eat our cake simultaneously here, but it will require some experimentation. In general I think it is better to have a more complex internal implementation if it improves error messages. I don't think we need to worry about compile times for the conversions, because since they are non-recursive, each conversion is approximately constant time in the size of the number.

**Comment by sdleffler** [comment]
Defer which to last, the `Continue` changes? Point taken on the compile times.

**Comment by kwf** [comment]
That's right—defer the `Continue` change to last, since it will likely be the most tricky, and nothing else depends on it.

**Comment by sdleffler** [comment]
Well, I have the rest implemented for the most part, minus the syntactic changes for macros. Other things currently do depend on what `Continue` uses, such as the macro compiler which currently emits unary integers.

**Comment by sdleffler** [comment]
It was much easier to change `Continue` than I thought. Should have this PR together and ready for review in the next day or two.
